docs/enhancements/fep-002-config-hooks.md

```diff
@@ -0,0 +1,204 @@
+
+### Non-Goals
+
+* Define a specific product integration, this is a generic mechanism
+  that could be used with any external system.
+
+## Proposal
+
+We define a specific type of configuration that can be used to query an external system.
+
+```yaml
+apiVersion: v1alpha1
+kind: Fleet
+metadata:
+  creationTimestamp: "2024-04-30T14:06:17Z"
+  generation: 1
+  labels: {}
+  name: optical-inspector-production-fleet
+  owner: ResourceSync/basic-nginx-r3sourcesync
+spec:
+  selector:
+    matchLabels:
+      device: optical-inspector
+  template:
+    metadata:
+      generation: 1
+    spec:
+      config:
+...
+...
+      - configType: ExternalConfigProviderSpec
+        externalEndpoint:
+          config-name: my-config
+          url: https://acm-cluster-flightctl-hook.example.com/flightctl-hooks/
+          httpCredentials: my-credentials
+...
+...
+
+      os:
+        image: quay.io/flightctl/flightctl-agent-basic-nginx:latest
+      monitoring:
+        systemd:
+          matchPatterns:
+          - microshift.service
+          - crio.service
+          - flightctl-agent.service
+status:
+  conditions:
+  - lastTransitionTime: "2024-05-09T14:20:21Z"
+    status: "False"
+    type: OverlappingSelectors
+```
+
+### High level format of the hook calls
+#### FlightCTL to external system
+
+##### CALL: POST https://acm-cluster-flightctl-hook.example.com/flightctl-hooks/
+[{"event: "ON_DEVICE_SPEC_UPDATE",
+ "event_uuid": "uuid-for-tracking",
+ "config_-_name": "my-config",
+ "device_metadata": { -json representation of the device metadata on our API, including the owner/fleet -},
+ "callback_endpoint": "https://api.flightctl.example.com/external-provider-callback/",
+ "callback_token": "my-token"}
+ "last_config_hash": "hash-of-the-last-applied-config"
+]
+
+###### RESPONSE: 200 OK
+ [{"event": "ON_DEVICE_SPEC_UPDATE",
+  "event_uuid": "uuid-for-tracking",
+  "device_name": "device-name",
+  "config-name": "my-config",
+  "configuration": { representation of the configuration to be injected on the device i.e. in ignition format },
+ }]
+
+###### RESPONSE: 304 Not Modified
+ [{"event": "ON_DEVICE_SPEC_UPDATE",
+  "event_uuid": "uuid-for-tracking",
+  "device_name": "device-name",
+  "config-name": "my-config",
+ }]
+
+ 
+#### External system to FlightCTL
+
+
+##### CALLS POST to https://api.flightctl.example.com/external-provider-callback/
+
+[{"event: "ON_CONFIG_UPDATE",
+ "event_uuid": "uuid-for-tracking",
+ "config_name": "my-config",
+ "device_names": [`device-name1`, `device-name2`],
+]
+
+
+### The hook events
+
+#### FlightCTL to external system
+* `ON_FLEET_JOIN`: This event is triggered when a device is added to a fleet for
+  the first time. The external system can use this event to provide the initial
+  configuration or to start building any specific configuration that needs to be
+  created.
+
+* `ON_DEVICE_SPEC_UPDATE`: This event is triggered when a device.spec is updated. At this
```

Comment:
This would be quite noisy, so we'd need to filter most of this out again. Again the notification could go to the Repository task who'd do the querying.

Reply:
Yes this would be unavoidably noisy. There is always the chance to optimize by grouping multiple requests.